File: src/component/parallax/ParallaxPanel.java
package component.parallax;

import java.util.ArrayList;
import java.util.Collections;

import org.newdawn.slick.Graphics;
import org.newdawn.slick.SlickException;
import org.newdawn.slick.geom.Vector2f;
import org.newdawn.slick.gui.GUIContext;

import component.Component;
import component.Panel;

/**
 * Panel that has a sorted list of {@code ParallaxSprite}.
 */
public class ParallaxPanel extends Panel {
	private ArrayList<ParallaxComponent> parallaxComponents;
	
	/**
	 * Construct a ParallaxPanel with a context, width, and height.
	 * @param context Context
	 * @param width Width
	 * @param height Height
	 */
	public ParallaxPanel(GUIContext context, int width, int height) {
		super(context, width, height);
		
		parallaxComponents = new ArrayList<ParallaxComponent>();
	}
	
	@Override
	public void render(GUIContext context, Graphics g) throws SlickException {
		super.render(container, g);
	}
	
	@Override
	public void update(int delta) {
		super.update(delta);
		
		for (int i = parallaxComponents.size() - 1; i >= 0; i--) {
			ParallaxComponent pc = parallaxComponents.get(i);
			pc.update(delta);
			
			if (pc.isExpired()) {
				parallaxComponents.remove(i);
				remove(pc);
			}
		}
	}
	
	@Override
	public void add(Component component, Vector2f location, ReferencePoint referencePoint, int xOffset, int yOffset) {
		if (!(component instanceof ParallaxComponent)) {
			return;
		}
		
		super.add(component, location, referencePoint, xOffset, yOffset);
		
		ParallaxComponent parallaxComponent = (ParallaxComponent) component;
		parallaxComponents.add(parallaxComponent);
		
		Collections.sort(parallaxComponents);
		
		for (int i = components.size() - 1; i >= 0; i--) {
			remove(components.get(i));
		}
		for (Component c : parallaxComponents) {
			components.add(c);
			c.setVisibleParent(this);
		}
	}
	
	/**
	 * Get the list of all the {@code ParallaxComponents}.
	 * @return List of Parallax COmponents
	 */
	public ArrayList<ParallaxComponent> getParallaxComponents() {
		return parallaxComponents;
	}
}

File: web/src/component/parallax/ParallaxPanel.ts
import { Panel } from '../Panel';
import { ParallaxComponent } from './ParallaxComponent';

/**
 * Container that manages multiple ParallaxComponents, updating them in distance-sorted order.
 * Components are sorted from farthest to nearest, so distant objects render first.
 */
export class ParallaxPanel extends Panel {
  private parallaxComponents: ParallaxComponent[] = [];
  private maxDistance: number = 100;
  private paused: boolean = false;
  private speed: number = 1;

  constructor(width: number, height: number) {
    super(width, height);
    this.setShouldUpdateComponents(true);
  }

  override update(delta: number): void {
    if (this.paused) {
      return;
    }

    for (let i = this.parallaxComponents.length - 1; i >= 0; i--) {
      const component = this.parallaxComponents[i];
      component.update(delta, this.speed);

      if (component.isExpired()) {
        this.removeParallaxComponent(component);
      }
    }

    super.update(delta);
  }

  addParallaxComponent(component: ParallaxComponent): void {
    this.parallaxComponents.push(component);
    this.sortComponents();

    this.components = [...this.parallaxComponents];
    for (const c of this.components) {
      c.setVisibleParent(this);
    }
  }

  removeParallaxComponent(component: ParallaxComponent): void {
    const index = this.parallaxComponents.indexOf(component);
    if (index > -1) {
      this.parallaxComponents.splice(index, 1);
      this.remove(component);
    }
  }

  private sortComponents(): void {
    this.parallaxComponents.sort((a, b) => b.getDistance() - a.getDistance());
  }

  setMaxDistance(maxDistance: number): void {
    this.maxDistance = maxDistance;
  }

  getMaxDistance(): number {
    return this.maxDistance;
  }

  setSpeed(speed: number): void {
    this.speed = speed;
  }

  getSpeed(): number {
    return this.speed;
  }

  pause(): void {
    this.paused = true;
    for (const component of this.parallaxComponents) {
      component.setPaused(true);
    }
  }

  resume(): void {
    this.paused = false;
    for (const component of this.parallaxComponents) {
      component.setPaused(false);
    }
  }

  isPaused(): boolean {
    return this.paused;
  }

  getParallaxComponents(): ParallaxComponent[] {
    return this.parallaxComponents;
  }
}
